barretenberg/ts/src/main.ts

```diff
@@ -50,33 +50,69 @@ async function computeCircuitSize(bytecodePath: string, honkRecursion: boolean,
+async function initUltraPlonk(bytecodePath: string, crsPath: string, subgroupSizeOverride = -1, honkRecursion = false) {
   const api = await Barretenberg.new({ threads });
 
-  const circuitSize = await getGates(bytecodePath, honkRecursion, api);
+  const circuitSize = await getGatesUltra(bytecodePath, honkRecursion, api);
   // TODO(https://github.com/AztecProtocol/barretenberg/issues/811): remove subgroupSizeOverride hack for goblin
   const subgroupSize = Math.max(subgroupSizeOverride, Math.pow(2, Math.ceil(Math.log2(circuitSize))));
-  if (subgroupSize > MAX_CIRCUIT_SIZE) {
-    throw new Error(`Circuit size of ${subgroupSize} exceeds max supported of ${MAX_CIRCUIT_SIZE}`);
-  }
 
+  if (subgroupSize > MAX_ULTRAPLONK_CIRCUIT_SIZE_IN_WASM) {
+    throw new Error(`Circuit size of ${subgroupSize} exceeds max supported of ${MAX_ULTRAPLONK_CIRCUIT_SIZE_IN_WASM}`);
+  }
   debug(`circuit size: ${circuitSize}`);
   debug(`subgroup size: ${subgroupSize}`);
   debug('loading crs...');
   // Plus 1 needed! (Move +1 into Crs?)
   const crs = await Crs.new(subgroupSize + 1, crsPath);
 
-  // Important to init slab allocator as first thing, to ensure maximum memory efficiency.
+  // Important to init slab allocator as first thing, to ensure maximum memory efficiency for Plonk.
   await api.commonInitSlabAllocator(subgroupSize);
 
   // Load CRS into wasm global CRS state.
   // TODO: Make RawBuffer be default behavior, and have a specific Vector type for when wanting length prefixed.
   await api.srsInitSrs(new RawBuffer(crs.getG1Data()), crs.numPoints, new RawBuffer(crs.getG2Data()));
-
   const acirComposer = await api.acirNewAcirComposer(subgroupSize);
   return { api, acirComposer, circuitSize, subgroupSize };
 }
 
+async function initUltraHonk(bytecodePath: string, crsPath: string, subgroupSizeOverride = -1) {
+  const api = await Barretenberg.new({ threads });
+
+  const circuitSize = await getGatesUltra(bytecodePath, /*honkRecursion=*/ true, api);
+  // TODO(https://github.com/AztecProtocol/barretenberg/issues/811): remove subgroupSizeOverride hack for goblin
+  const subgroupSize = Math.max(subgroupSizeOverride, Math.pow(2, Math.ceil(Math.log2(circuitSize))));
+
+  debug(`circuit size: ${circuitSize}`);
+  debug(`subgroup size: ${subgroupSize}`);
+  debug('loading crs...');
+  // Plus 1 needed! (Move +1 into Crs?)
+  const crs = await Crs.new(subgroupSize + 1, crsPath);
+
+  // Load CRS into wasm global CRS state.
+  // TODO: Make RawBuffer be default behavior, and have a specific Vector type for when wanting length prefixed.
+  await api.srsInitSrs(new RawBuffer(crs.getG1Data()), crs.numPoints, new RawBuffer(crs.getG2Data()));
+  return { api, circuitSize, subgroupSize };
+}
+
+async function initClientIVC(bytecodePath: string, crsPath: string) {
+  const api = await Barretenberg.new({ threads });
+
+  debug('loading BN254 and Grumpkin crs...');
+  // Plus 1 needed! (Move +1 into Crs?)
+  const crs = await Crs.new(2 ** 18 + 1, crsPath);
```

Comment:
we're going to need to figure out how to count gates for ClientIVC. It will probably just be a megaGates function, but I didn't create it now because it doesn't make sense with the current ClientIVC (which injects more gates from folding and merge verifiers during runtime).